Makefile

```diff
@@ -47,55 +59,48 @@ SCRIPTS = \
 	watchdog-report.sh
 
 SHELL_CMD=bash
-define BUILD_CONTAINER
-FROM golang:1.9.1-alpine
-RUN apk add --no-cache openssh-client git gcc linux-headers libc-dev util-linux libpcap-dev bash vim make
-RUN deluser $(USER) ; delgroup $(GROUP) || :
-RUN sed -ie /:$(UID):/d /etc/passwd /etc/shadow ; sed -ie /:$(GID):/d /etc/group || :
-RUN addgroup -g $(GID) $(GROUP) && adduser -h /home/$(USER) -G $(GROUP) -D -H -u $(UID) $(USER)
-ENV HOME /home/$(USER)
-endef
 
-.PHONY: all clean vendor
+.PHONY: all clean vendor $(GOBUILDER) builder-image build build-docker versioninfo
 
 all: obj build
 
 obj:
-	@rm -rf $(BINDIR)
-	@mkdir -p $(BINDIR)
-
-build:
-	@echo Building version $(BUILD_VERSION)
-	@mkdir -p var/tmp/zededa
-	@echo $(BUILD_VERSION) >$(BINDIR)/versioninfo
-	@for app in $(APPS); do \
-		echo $$app; \
-		CGO_ENABLED=0 \
-		GOOS=linux \
-		GOARCH=$(ARCH) $(GOBUILD) \
-			-ldflags -X=main.Version=$(BUILD_VERSION) \
-			-o $(BINDIR)/$$app github.com/zededa/go-provision/$$app || exit 1; \
-	done
-	@for app in $(APPS1); do \
-		echo $$app; \
-		rm -f $(BINDIR)/$$app; \
-		ln -s $(APPS) $(BINDIR)/$$app; \
-	done
+	@rm -rf $(DISTDIR)
+	@mkdir -p $(DISTDIR)
+
+$(DISTDIR): 
+	mkdir -p $(DISTDIR)
+
+versioninfo:
+	@echo $(BUILD_VERSION) >$(DISTDIR)/versioninfo
+
+build: $(APPS) $(APPS1)
+
+$(APPS): $(DISTDIR) builder-image
+	@echo "Building $@"
+	@$(GOBUILD) -ldflags -X=main.Version=$(BUILD_VERSION) \
+		-o $(DISTDIR)/$@ github.com/zededa/go-provision/$@
+
+$(APPS1): $(DISTDIR)
+	@echo $@
+	@rm -f $(DISTDIR)/$@
+	@ln -s $(APPS) $(DISTDIR)/$@
+
+shell: builder-image
+	@$(DOCKER_GO) $(SHELL_CMD) 
 
 build-docker:
 	docker build $(DOCKER_ARGS) -t $(DOCKER_TAG) .
 
 build-docker-git:
 	git archive HEAD | docker build $(DOCKER_ARGS) -t $(DOCKER_TAG) -
 
-export BUILD_CONTAINER
-eve-build-$(USER):
-	@echo "$$BUILD_CONTAINER" | docker build -t $@ - >/dev/null
-
-shell: eve-build-$(USER)
-	@docker run -it --rm -u $(USER) -w /home/$(USER) \
-	  -v $(CURDIR)/.go:/go -v $(CURDIR):/go/src/$(GOMODULE) -v $${HOME}:/home/$(USER) \
-	$< $(SHELL_CMD)
+builder-image: $(GOBUILDER)
```

Comment:
@deitch this may be just me (and I'd like to get @eriknordmark and @kalyan-nidumolu and more specifically @aw-zededa opinion here) but I tend to treat anything that has a Dockerfile as being a non-transient container (e.g. a container that has to end up in Docker HUB).

These containers (build-eve-USERNAME) NEVER  supposed to end up in a Docker HUB (in fact if they do -- it'll be a bad thing(tm)).

Hence my strong preference will be to keep the mechanics transient -- with the Dockerfile being piped into docker on the fly, rather than introducing builder/Dockerfile as you do below.

Reply:
>  I tend to treat anything that has a Dockerfile as being a non-transient container

I haven't seen it that way. There are plenty of cases were dockerfiles are used for testing containers, interim containers, etc. It becomes messy and impractical (and hard to understand diffs) when it is embedded in something else.

> These containers (build-eve-USERNAME) NEVER supposed to end up in a Docker HUB (in fact if they do -- it'll be a bad thing(tm)).

Agreed 100%. But the determination for its ability to go somewhere or not isn't the existence of the Dockerfile, but rather whether the container has a name and tags. This one does - `build-eve-$(USERNAME):latest` (with the `latest` implicit), so in theory it can go out, whether you build it via Dockerfile or inline.

What _does_ prevent it is giving it a name that there is almost no way you _could_ push it out without trying really hard. The name `build-eve-rvs` is really `library/build-eve-rvs:latest` which we can rely on with reasonable certainty is something you cannot push out.

I usually agree with you on most of these things, but if we have a container image to build, let's do it in a Dockerfile and name it something that we cannot leak it (which you did).

Reply:
> > I tend to treat anything that has a Dockerfile as being a non-transient container
> 
> I haven't seen it that way. There are plenty of cases were dockerfiles are used for testing containers, interim containers, etc. It becomes messy and impractical (and hard to understand diffs) when it is embedded in something else.

+1  This is especially important for linting.

Reply:
OK. Agreed @deitch and @aw-zededa -- you've convinced me ;-) But! Lets at least call the top level folder containing it something super explicit.  dev-support is a good name that @aw-zededa and I are used to in Apache world ;-)

@deitch can you please make that change plus Go version and resubmit?

Reply:
Done